# Современные методы анализа данных и машинного обучения, БИ

## НИУ ВШЭ, 2024-25 учебный год

### Домашнее задание №1. Библиотеки Numpy & Pandas

Задание выполнил(а):

   Климов Дмитрий ББИ226

### Общая информация

__Дата выдачи:__ 19.01.2025

__Дедлайн:__ 04:00 29.01.2025

### Оценивание и штрафы

Каждая задача первого раздела (Numpy) оценивается в 2 балла.

Оценка за первый раздел вычисляется по следующей формуле:

$$
s_{\text{numpy}} \times 1/2 ,
$$

где $s_{\text{numpy}}$  — количество баллов, которое вы набрали в сумме по задачам первого раздела.

Каждая задача второго раздела (Pandas) оценивается в 3 балла.

Оценка за второй раздел вычисляется по следующей формуле:

$$
s_{\text{pandas}} \times 10/54 ,
$$

где $s_{\text{pandas}}$  — количество баллов, которое вы набрали в сумме по задачам второго раздела.

Общая оценка за ДЗ вычисляется по следующей формуле:

$$
O_{1} \times 3/10 + O_2 \times 7/10 ,
$$

где $O_1, O_2$  — оценки за первый и второй раздел соответственно.

За сдачу задания позже срока на итоговую оценку за задание накладывается штраф в размере 1 **вторичный** балл в день, но  задержка не может быть больше недели.

**Важно:** В этом домашнем задании предусмотрено несколько заданий, за которые можно получить бонусные баллы. Бонусные баллы прибавляются к обычным баллам, и может выйти так, что ваша итоговая оценка станет больше 10. В таком случае именно это и будет являться вашей оценкой. То есть за это ДЗ **можно** получить больше 10

__Внимание!__ Домашнее задание выполняется самостоятельно. Не допускается помощь в решении домашнего задания от однокурсников или третьих лиц. «Похожие» решения считаются плагиатом, и все задействованные студенты — в том числе и те, у кого списали, — не могут получить за него больше 0 баллов. Использование в решении домашнего задания генеративных моделей (ChatGPT и так далее) за рамками справочной и образовательной информации для генерации кода задания — считается плагиатом, и такое домашнее задание оценивается в 0 баллов.

### Формат сдачи

Загрузка файлов с решениями происходит в системе [Anytask](https://anytask.org/). Необходимо загружать файл с расширением .ipynb (питоновский ноутбук)

Инвайт для подключения к странице курса в системе — будет указан для каждой группы в нашем телеграмм-канале позднее.

## Раздел 1. Numpy

### О задании

В данном разделе вы попрактикуетесь в работе с библиотекой numpy!

In [188]:
import numpy as np

Во всех задачах необходимо написать код решения внутри функции и проверить его работу, вызвав функцию для данных из условия.

При решении задач запрещается использовать циклы (`for`, `while`) и оператор `if`.

Везде, где встречаются массивы или матрицы, подразумевается, что это `numpy.array`.

### 1

Напишите функцию, возвращающую округленную взвешенную сумму оценок по данным оценкам и весам. Можете посчитать свою оценку за курс :) В нашем случае вес экзамена равен 0.25, вес домашек - 0.1, вес групповых проектов - 0.2, вес аудиторки - 0.2, вес контрольной 1 - 0.1, вес контрольной 2 - 0.1 и вес проекта - 0.05. Например, если экзамен написан на 6, домашки сданы на 9, проекты закрыты на 8, оценка за аудиторку 7, контрольная 1 написана на 10, контрольная 2 написана на 7 и проект сдан на 10 - итоговая оценка будет хорошо - 8

In [189]:
def result_mark(weights: np.array, marks: np.array) -> int:
    
    # проверка на то, что сумма весов равна 1
    if not np.isclose(np.sum(weights), 1.0):
        raise ValueError("Сумма весов должна равняться 1")
    
    # проверка на то, что элементов контроля столько, сколько весов
    if weights.shape != marks.shape:
        raise ValueError("Весов и оценок должно быть одинаковое количество")

    final_sum = np.dot(marks, weights)
    return round(final_sum)

In [190]:
weights = np.array([0.25, 0.1, 0.2, 0.2, 0.1, 0.1, 0.05])
marks = np.array([6, 9, 8, 7, 10, 7, 10])
result_mark(weights, marks)

8

### 2

Напишите функцию, меняющую каждое третье (начиная с 0) значение массива целых чисел на заданное число. Например, если на вход поступает массив `array([3, 5, 1, 0, -3, 22, 213436])` и число `-111`, то на выходе должен получиться массив `array([-111, 5, 1, -111, -3, 22, -111])`.

In [191]:
def change_array(array: np.array, number: int) -> np.array:
    array[::3] = number
    return array

In [192]:
array = np.array([3, 5, 1, 0, -3, 22, 213436])
number = -111
change_array(array, number)

array([-111,    5,    1, -111,   -3,   22, -111])

### 3

Напишите функцию, выдающую индексы «близких» элементов заданных массивов, а именно тех пар элементов, чей модуль разницы не превосходит заданного значения. Например, если на вход поступают массив `array([1.5, 0.5, 2, -4.1, -3, 6, -1])`, массив `array([1.2, 0.5, 1, -4, 3, 0, -1.2])` и число `0.5`, то на выходе должен получиться массив `array([0, 1, 3, 6])` (_**важно: не `tuple`, а одномерный массив типа `numpy.ndarray` (то есть `.ndim` от него равно 1)!**_).

In [193]:
def find_close(array1: np.array, array2: np.array,
               precision: float) -> np.array:
    diff = np.abs(array1-array2)
    res = np.where(diff <= precision)[0]
    return res

In [194]:
array1 = np.array([1.5, 0.5, 2, -4.1, -3, 6, -1])
array2 = np.array([1.2, 0.5, 1, -4, 3, 0, -1.2])
precision = 0.5
find_close(array1, array2, precision)

array([0, 1, 3, 6])

### 4

Напишите функцию, которая составляет блочную матрицу из четырех блоков, где каждый блок - это заданная матрица. Например, если на вход поступает матрица
$$
\begin{pmatrix}
0 & 1 & 2\\
3 & 4 & 5\\
\end{pmatrix},
$$
то ответом будет матрица
$$
\begin{pmatrix}
0 & 1 & 2 & 0 & 1 & 2\\
3 & 4 & 5 & 3 & 4 & 5\\
0 & 1 & 2 & 0 & 1 & 2\\
3 & 4 & 5 & 3 & 4 & 5\\
\end{pmatrix}
$$

In [195]:
def block_matrix(block: np.array) -> np.array:
    return np.block([[block, block], [block, block]])

In [196]:
block = np.array([[0, 1, 2], [3, 4, 5]])
block_matrix(block)

array([[0, 1, 2, 0, 1, 2],
       [3, 4, 5, 3, 4, 5],
       [0, 1, 2, 0, 1, 2],
       [3, 4, 5, 3, 4, 5]])

### 5

Напишите функцию, вычисляющую произведение всех ненулевых диагональных элементов на главной диагонали данной квадратной матрицы. Например, если на вход поступает матрица
$$
\begin{pmatrix}
0 & 1 & 2\\
3 & 4 & 5\\
6 & 7 & 8\\
\end{pmatrix},
$$
то ответом будет 32.

In [197]:
def diag_prod(matrix: np.array) -> int:
    els = np.diagonal(matrix)
    cond_els = els[els != 0]
    return np.prod(cond_els)

In [198]:
matrix = np.array([[0, 1, 2], [3, 4, 5], [6, 7, 8]])
diag_prod(matrix)

np.int64(32)

### 6

Для улучшения качества работы некоторых алгоритмов машинного обучения может быть полезно использовать [нормализацию данных](https://vk.cc/8xmfQk), чтобы привести признаки в выборке к одному масштабу — а именно, из каждого столбца вычесть среднее его значений и поделить на их стандартное отклонение. Напишите функцию, нормализующую входящую матрицу (по столбцам). Например, если на вход подается матрица
$$
\begin{pmatrix}
1 & 4 & 4200\\
0 & 10 & 5000\\
1 & 2 & 1000\\
\end{pmatrix},
$$
то результатом с точностью до сотых будет матрица
$$
\begin{pmatrix}
0.71 & -0.39 & 0.46\\
-1.41 & 1.37 & 0.93\\
0.71 & -0.98 & -1.39\\
\end{pmatrix}
$$

_Подсказка. Учтите, что в вашей матрице не должно получаться никаких nan. Подумайте, в каком случае они могут возникнуть и как обойти эту проблему._


In [199]:
def normalize(matrix: np.array) -> np.array:
    std = np.std(matrix, axis=0)
    std_wout_zeros = np.where(std == 0, 1, std)
    return np.round((matrix - np.mean(matrix, axis = 0)) / std_wout_zeros, 2)

In [200]:
matrix = np.array([[1, 4, 4200], [0, 10, 5000], [1, 2, 1000]])
normalize(matrix)

array([[ 0.71, -0.39,  0.46],
       [-1.41,  1.37,  0.93],
       [ 0.71, -0.98, -1.39]])

nan-значения могут быть в том случае если все значения в столбце равны 0, тогда и стандартное отклонение будет равно 0, а результатом деления окажется nan, поэтому нам стоит задать значение для всех nan элементов

### 7

Напишите функцию, вычисляющую какую-нибудь первообразную данного полинома (в качестве константы возьмите ваше любимое число). Например, если на вход поступает массив коэффициентов `array([4, 6, 0, 1])`, что соответствует полиному $4x^3 + 6x^2 + 1$, на выходе получается массив коэффициентов `array([1, 2, 0, 1, -2])`, соответствующий полиному $x^4 + 2x^3 + x - 2$.

In [201]:
def antiderivative(coefs: np.array) -> np.array:
    n = len(coefs)
    new_coefs = np.zeros(n + 1, dtype=float)

    for i in range(n):
        deg = n - 1 - i
        new_coefs[i] = coefs[i] / (deg + 1)

    new_coefs[-1] = 11
    return new_coefs

In [202]:
coefs = np.array([4, 6, 0, 1])
antiderivative(coefs)

array([ 1.,  2.,  0.,  1., 11.])

в качестве dtype взял float, тк существуют случаи, когда коэффициенты первообразной будут получаться не целыми, думаю, это более правильное решение)

### 8

Напишите функцию, делающую данную [треугольную матрицу](https://ru.wikipedia.org/wiki/%D0%A2%D1%80%D0%B5%D1%83%D0%B3%D0%BE%D0%BB%D1%8C%D0%BD%D0%B0%D1%8F_%D0%BC%D0%B0%D1%82%D1%80%D0%B8%D1%86%D0%B0) симметричной. Например, если на вход поступает матрица
$$
\begin{pmatrix}
1 & 2 & 3 & 4\\
0 & 5 & 6 & 7\\
0 & 0 & 8 & 9\\
0 & 0 & 0 & 10\\
\end{pmatrix},
$$
то на выходе должна быть матрица
$$
\begin{pmatrix}
1 & 2 & 3 & 4\\
2 & 5 & 6 & 7\\
3 & 6 & 8 & 9\\
4 & 7 & 9 & 10\\
\end{pmatrix}.
$$

In [203]:
def make_symmetric(matrix: np.array) -> np.array:
    return matrix + matrix.T - np.diag(matrix.diagonal())

In [204]:
matrix = np.array([[1, 2, 3, 4], [0, 5, 6, 7], [0, 0, 8, 9], [0, 0, 0, 10]])
make_symmetric(matrix)

array([[ 1,  2,  3,  4],
       [ 2,  5,  6,  7],
       [ 3,  6,  8,  9],
       [ 4,  7,  9, 10]])

### 9

Напишите функцию, создающую прямоугольную матрицу из m одинаковых строк, заполненных последовательными натуральными числами от a до b включительно в возрастающем порядке. Например, если m = 5, a = 3, b = 10, то на выходе будет матрица
$$
\begin{pmatrix}
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
\end{pmatrix}
$$

In [205]:
def construct_matrix(m: int, a: int, b: int) -> np.array:
    str = np.arange(a, b+1)
    return np.tile(str, (m,1))

In [206]:
m = 5
a = 3
b = 10
construct_matrix(m, a, b)

array([[ 3,  4,  5,  6,  7,  8,  9, 10],
       [ 3,  4,  5,  6,  7,  8,  9, 10],
       [ 3,  4,  5,  6,  7,  8,  9, 10],
       [ 3,  4,  5,  6,  7,  8,  9, 10],
       [ 3,  4,  5,  6,  7,  8,  9, 10]])

### 10

Напишите функцию, вычисляющую [косинусную близость](https://en.wikipedia.org/wiki/Cosine_similarity) двух векторов. Например, если на вход поступают вектора `array([-2, 1, 0, -5, 4, 3, -3])` и `array([0, 2, -2, 10, 6, 0, 0])`, ответом будет -0.25.

In [207]:
def cosine_similarity(vec1: np.array, vec2: np.array) -> float:
    skal = np.dot(vec1, vec2)
    norm1 = np.linalg.norm(vec1)
    norm2 = np.linalg.norm(vec2)
    return skal / (norm1 * norm2)

In [208]:
vec1 = np.array([-2, 1, 0, -5, 4, 3, -3])
vec2 = np.array([0, 2, -2, 10, 6, 0, 0])
cosine_similarity(vec1, vec2)

np.float64(-0.25)

## Раздел 2. Pandas

### О задании

В данном разделе вы попрактикуетесь в работе с библиотекой pandas!

In [209]:
import pandas as pd

Обратите внимание, что практически во всех заданиях вам потребуется написать какой-то текст в виде вывода/результата/итога/какого-то резюме. Пожалуйста, не ленитесь, и пишите выводы ***подробно***, существенная часть баллов будет выставляться именно за выводы, так как в анализе данных они имеют большое значение. Таким образом, на любое задание, сформулированное в виде вопроса, отвечайте текстом, подкрепляя ваши слова написанным и выполненным кодом. Текст выводов/комментариев пишите в отдельных ячейках, используя средства текстового редактора Jupyter Notebook/Google Colab, при необходимости используйте средства *выделения*.

При решении задач, когда вам необходимо сделать какой-то фильтр, срез, выборку, или иное действие с данными, запрещается использовать циклы (`for`, `while`) и оператор `if`.


### 1

В этом разделе домашнего задания вы поработаете с данными исследования зависимости средней продолжительности жизни по странам от количества употребляемого в стране алкоголя.

Вашему вниманию представлено два датасета: drinks.csv и lifeexpectancy-verbose.csv

Первая таблица содержит в себе сводные данные по количеству употребляемого алкоголя в каждой стране.

Описание датасета:
- country -	Name of the country (String)
- beer_servings	- Number of beer servings per person in the country (Integer)
- spirit_servings	- Number of spirit servings per person in the country (Integer)
- wine_servings	- Number of wine servings per person in the country(Integer)
- total_litres_of_pure_alcohol -	Total litres of pure alcohol consumed per person in the country (Float)

Вторая таблица содержит в себе информацию о научных статьях и исследованиях о продолжительности жизни (в разных категориях и для разных групп) по той или иной стране.

Откройте в ноутбуке оба файла с данными. Какой размер имеют таблицы? Выведите первые 5 и последние 5 строк для обеих таблиц. Сколько уникальных стран представлено в каждой из таблиц? Сделайте выводы.

In [210]:
drinks = pd.read_csv('drinks.csv')
lifeexp = pd.read_csv('lifeexpectancy-verbose.csv')
print(drinks.shape, lifeexp.shape)

(193, 6) (6408, 17)


Видим, что таблица с алкоголем содержит 193 наблюдения (пока что предположительно по 193 странам) по 6 признакам, а таблица с продолжительностью жизни содержит 6408 наблюдений по 17 признакам

In [211]:
country_mapping = {
    'Iran (Islamic Republic of)': 'Iran',
    'Republic of Korea': 'South Korea',
    'Syrian Arab Republic': 'Syria',
    'Lao People\'s Democratic Republic': 'Laos',
    'Democratic People\'s Republic of Korea': 'North Korea',
    'United Republic of Tanzania': 'Tanzania',
    'Republic of Moldova': 'Moldova',
    'Viet Nam': 'Vietnam',
    'Venezuela (Bolivarian Republic of)': 'Venezuela',
    'United Kingdom of Great Britain and Northern Ireland': 'United Kingdom',
    'Micronesia (Federated States of)': 'Micronesia',
    'Trinidad and Tobago': 'Trinidad & Tobago',
    'Democratic Republic of the Congo': 'DR Congo',
    'Brunei Darussalam': 'Brunei',
    'Saint Vincent and the Grenadines': 'St. Vincent & the Grenadines',
    'Timor_Leste': 'Timor-Leste',
    'United States of America': 'USA',
    'Côte d\'Ivoire': 'Cote d\'Ivoire',
    'Bolivia (Plurinational State of)': 'Bolivia',
    'The former Yugoslav republic of Macedonia': 'Macedonia',
    'Saint Lucia': 'St. Lucia',
    'Saint Kitts and Nevis': 'St. Kitts & Nevis',
    'Antigua and Barbuda': 'Antigua & Barbuda',
    'Bosnia and Herzegovina': 'Bosnia-Herzegovina',
    'Guinea-Bissau': 'Guinea_Bissau',
    'Sao Tome and Principe': 'Sao Tome & Principe'
}

lifeexp['CountryDisplay'] = lifeexp['CountryDisplay'].replace(country_mapping)
drinks['country'] = drinks['country'].replace(country_mapping)


In [212]:
drinks.head()

index      country  beer_servings  spirit_servings  wine_servings  \
0      0  Afghanistan              0                0              0   
1      1      Albania             89              132             54   
2      2      Algeria             25                0             14   
3      3      Andorra            245              138            312   
4      4       Angola            217               57             45   

   total_litres_of_pure_alcohol  
0                           0.0  
1                           4.9  
2                           0.7  
3                          12.4  
4                           5.9

In [213]:
drinks.tail()

index    country  beer_servings  spirit_servings  wine_servings  \
188    188  Venezuela            333              100              3   
189    189    Vietnam            111                2              1   
190    190      Yemen              6                0              0   
191    191     Zambia             32               19              4   
192    192   Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol  
188                           7.7  
189                           2.0  
190                           0.1  
191                           2.5  
192                           4.7

Можем сделать вывод о том, что существенными тут являются 5 признаков (тк один из них - индекс - не несет никакой полезной информации)

In [214]:
lifeexp.head()

index        GhoCode                                       GhoDisplay  \
0      0  WHOSIS_000001                 Life expectancy at birth (years)   
1      1  WHOSIS_000002  Healthy life expectancy (HALE) at birth (years)   
2      2  WHOSIS_000001                 Life expectancy at birth (years)   
3      3  WHOSIS_000001                 Life expectancy at birth (years)   
4      4  WHOSIS_000015                Life expectancy at age 60 (years)   

  PublishStateCode PublishStateDisplay  YearCode  YearDisplay RegionCode  \
0        PUBLISHED           Published      1990         1990        AMR   
1        PUBLISHED           Published      2012         2012        EUR   
2        PUBLISHED           Published      2000         2000        EMR   
3        PUBLISHED           Published      2000         2000        AFR   
4        PUBLISHED           Published      1990         1990       SEAR   

           RegionDisplay WorldBankIncomeGroupGroupCode  \
0               Americas                        WB_LMI   
1                 Europe                         WB_HI   
2  Eastern Mediterranean                         WB_LI   
3                 Africa                        WB_LMI   
4        South_East Asia                        WB_LMI   

  WorldBankIncomeGroupDisplay CountryCode CountryDisplay SexCode  SexDisplay  \
0         Lower_middle_income         NIC      Nicaragua     MLE        Male   
1                 High_income         IRL        Ireland     MLE        Male   
2                  Low_income         YEM          Yemen    BTSX  Both sexes   
3         Lower_middle_income         NGA        Nigeria    FMLE      Female   
4         Lower_middle_income         THA       Thailand    FMLE      Female   

   DisplayValue  Numeric  
0            68     68.0  
1            69     69.3  
2            61     61.0  
3            48     48.0  
4            19     19.0

In [215]:
lifeexp.tail()

index        GhoCode                                       GhoDisplay  \
6403   6403  WHOSIS_000001                 Life expectancy at birth (years)   
6404   6404  WHOSIS_000015                Life expectancy at age 60 (years)   
6405   6405  WHOSIS_000001                 Life expectancy at birth (years)   
6406   6406  WHOSIS_000002  Healthy life expectancy (HALE) at birth (years)   
6407   6407  WHOSIS_000001                 Life expectancy at birth (years)   

     PublishStateCode PublishStateDisplay  YearCode  YearDisplay RegionCode  \
6403        PUBLISHED           Published      2000         2000        EMR   
6404        PUBLISHED           Published      1990         1990        WPR   
6405        PUBLISHED           Published      1990         1990        AFR   
6406        PUBLISHED           Published      2012         2012        EUR   
6407        PUBLISHED           Published      2013         2013        EUR   

              RegionDisplay WorldBankIncomeGroupGroupCode  \
6403  Eastern Mediterranean                        WB_LMI   
6404        Western Pacific                        WB_UMI   
6405                 Africa                         WB_LI   
6406                 Europe                        WB_LMI   
6407                 Europe                           NaN   

     WorldBankIncomeGroupDisplay CountryCode CountryDisplay SexCode  \
6403         Lower_middle_income         MAR        Morocco    BTSX   
6404         Upper_middle_income         PLW          Palau    BTSX   
6405                  Low_income         BFA   Burkina Faso    FMLE   
6406         Lower_middle_income         ALB        Albania     MLE   
6407                         NaN         ARM        Armenia    BTSX   

      SexDisplay  DisplayValue   Numeric  
6403  Both sexes            68  68.00000  
6404  Both sexes            16  16.00000  
6405      Female            51  51.00000  
6406        Male            64  63.50000  
6407  Both sexes            71  71.02699

Видим, что из 17 признаков существенных сильно меньше: тут есть индекс, и для качественных признаков созданы еще дополнительно коды - сокращенные названия

In [216]:
drinks['country'].nunique()

193

вспоминая, что в напитках всего 193 строки, можно убедиться в том, что для каждой страны в данном датасете ровно одно наблюдение

In [217]:
lifeexp['CountryDisplay'].nunique()

194

### 2

Для каждого признака в рассматриваемых данных укажите вид этого признака и кратко обоснуйте, почему вид именно такой.

## lifeexpectancy
- GhoDisplay: Категориальный. Текстовые данные, классифицирующие тип статьи
- YearDisplay: Вещественный. Не временной ряд, так как в данных нет четкой временной линии от определенного года к определенному году, то есть нет никакой тенденции.
- RegionDisplay: Категориальный. Континент
- CountryDisplay: Категориальный. Страна
- SexDisplay: Категориальный.Не бинарный, так как, по мимо двух полов, есть еще строки, в которых встречается "Both sexes".
- Numeric: Вещественный. Продолжительность жизни

## drinks
- country: Категориальный. Текстовое наименование страны
- beer_servings: Числовой (int). Количество порций пива на человека
- spirit_servings: Числовой (int). Количество порций крепкого алкоголя на человека
- wine_servings: Числовой (int). Количество порций вина на человека
- total_litres_of_pure_alcohol: Числовой (int). Суммарное количество литров алкоголя на человека

### 3

Внимательно изучите признаки обеих таблиц. Как вы считаете, какая из таблиц является главной (а какая сопутствующей)? Почему вы так думаете?

Какой признак может представлять существенный интерес для исследования и являться целевой переменной в рамках представленного датасета? Ответ обоснуйте.

***Бонус (1 бон. балл):*** Попробуйте описать задачу машинного обучения, которую вы могли бы решать для данного датасета с учетом выбранной вами раннее целевой переменной и проблематики исследования. Подробнее про постановку и формализацию задач машинного обучения, а также про различные типы таких задач — вы можете прочитать, например, вот в этой [статье](https://habr.com/ru/company/simbirsoft/blog/565396/) (в начале)

- Главной является таблица, содержащая данные о продолжительности жизни, так как в таблице, содержащей данные о потреблении алкоголя каждая строка уникальна и соответствует определенной стране (то есть в ней уже есть аггрегированные данные по странам).А в таблице, содержащей данные о продолжительности жизни, по каждой стране есть разная информация из разных статей и о разных полах.
- Целевой переменной в данном датасете является переменная "Numeric", которая отображает данные о продолжительности жизни в разных странах в разное время, так как именно по этому признаку будет определяться зависимость от объемов потребляемого алкоголя.
- Задачу МО для данной целевой переменной можно описать примерно так: прогнозирование продолжительности жизни на основе данных из двух таблиц: beer_servings, spirit_servings, wine_servings, total_litres_of_pure_alcohol, CountryDisplay и SexDisplay. Это будет являться задачей регрессии (тк предиктим число).

### 4

Какова доля статей о продолжительности жизни в Африке? Какова доля статей о продолжительности жизни мужчин и доля статей о продолжительности жизни женщин среди всех статей по Африке?

***Бонус (2 бон. балла):*** Сравните эти доли с аналогичными долями по другим регионам. Как вы можете объяснить возникающие сходства и различия? Прокомментируйте полученные результаты.

In [218]:
lifeexp.RegionDisplay.value_counts() / 6408 * 100

RegionDisplay
Europe                   27.294007
Africa                   23.782772
Americas                 18.024345
Western Pacific          13.904494
Eastern Mediterranean    11.329588
South_East Asia           5.664794
Name: count, dtype: float64

Доля статей о продолжительности жизни в Африке составляет примерно 24%

In [219]:
lifeexp[lifeexp['RegionDisplay'] == 'Africa']['SexDisplay'].value_counts()

SexDisplay
Female        508
Both sexes    508
Male          508
Name: count, dtype: int64

Как мы можем видеть, в Африке доли статей о мужчинах и женщинах равны между собой, также еще есть доля статей, в которых упоминаются оба пола, и она такая же, как у статей только про мужчин или женщин. В связи с этим, можно сказать, что доли статей о мужчинах и женщинах по отдельности равны **33.3%**

In [220]:
lifeexp.pivot_table(index='RegionDisplay', columns='SexDisplay', aggfunc='size')

SexDisplay             Both sexes  Female  Male
RegionDisplay                                  
Africa                        508     508   508
Americas                      385     385   385
Eastern Mediterranean         242     242   242
Europe                        583     583   583
South_East Asia               121     121   121
Western Pacific               297     297   297

Такую же ситуацию, как с Африкой, мы наблюдаем и с другими частями света, везде доли статей распределены одинаково и равны 33.3% от общего кол-ва статей внутри региона.
Мы наблюдаем одинаковое кол-во статей о женщинах, о мужчинах и об обоих полах в совокупности, поскольку по стандартам сбора данных необходимо равномерное распределение внимания к продолжительности жизни обоих полов в рамках региона.
Различия в виде разного объема данных, собранных в разных регионах, может объясняться доступностью информации в этих регионах. Совершенно логично, что данных по Европе собрано больше, чем по, к примеру, Южно-Восточной Азии, так как в Европе гораздо больше развита медицина и доступность данных.

### 5

Сколько всего существует видов статей с точки зрения оцениваемого параметра продолжительности жизни? Какой из этих видов статей самый популярный (написано наибольшее количество научных работ)? В рамках найденного вами самого популярного вида статей проведите сравнение, относительно какой из половых групп: мужчины, женщины или же оба пола — статей в нем было больше всего?

In [221]:
lifeexp.GhoDisplay.unique()

array(['Life expectancy at birth (years)',
       'Healthy life expectancy (HALE) at birth (years)',
       'Life expectancy at age 60 (years)'], dtype=object)

В датасете есть всего 3 вида статей: Life expectancy at birth (years), Healthy life expectancy (HALE) at birth (years), Life expectancy at age 60 (years)

In [222]:
lifeexp.GhoDisplay.value_counts()

GhoDisplay
Life expectancy at birth (years)                   2328
Life expectancy at age 60 (years)                  2328
Healthy life expectancy (HALE) at birth (years)    1752
Name: count, dtype: int64

Мы видим, что наиболее популярных видов статей два: Life expectancy at birth (years), Life expectancy at age 60 (years), по каждому из них написано по 2328 статей.

In [223]:
lifeexp[lifeexp['GhoDisplay'] == 'Life expectancy at birth (years)']['SexDisplay'].value_counts()

SexDisplay
Male          776
Both sexes    776
Female        776
Name: count, dtype: int64

In [224]:
lifeexp[lifeexp['GhoDisplay'] == 'Life expectancy at age 60 (years)']['SexDisplay'].value_counts()

SexDisplay
Female        776
Both sexes    776
Male          776
Name: count, dtype: int64

Мы можем заметить, что в разрезе самых популярных видов статей также написано одинаковое кол-во статей про мужчин, женщин и про оба пола в совокупности, что еще раз подтверждает мою гипотезу, выдвинутую в задании 4

### 6

Выведите минимальную, среднюю, медианную и максимальную оцененную продолжительность жизни людей от рождения (life expectancy at birth). Выберите любые три страны на ваш вкус и проделайте то же самое только для статей по этим странам. Что можно сказать на основе результатов, которые вы получили?

In [225]:
lifeexp[lifeexp['GhoDisplay'] == 'Life expectancy at birth (years)']['Numeric'].describe()

count    2328.000000
mean       68.441379
std        10.051907
min        38.000000
25%        62.000000
50%        70.857895
75%        76.000000
max        87.154570
Name: Numeric, dtype: float64

Среди статей и продолжительности жизни людей от рождения выявлены такие показатели, как:
- Минимальная продолжительность жизни = 38 лет
- Средняя продолжительность жизни = 68.44 лет
- Медианная продолжительность жизни = 70.86 лет
- Максимальная продолжительность жизни = 87 лет

In [226]:
lifeexp[(lifeexp['CountryDisplay'] == 'Germany') &
     (lifeexp['GhoDisplay'] == 'Life expectancy at birth (years)')].groupby('CountryDisplay')['Numeric'].agg(['min', 'mean', 'median', 'max'])

min       mean    median       max
CountryDisplay                                     
Germany         72.0  78.814174  78.77455  83.27878

Среди статей и продолжительности жизни людей от рождения **в Германии** выявлены такие показатели, как:

- Минимальная продолжительность жизни = 72 года
- Средняя продолжительность жизни = 78.81 лет
- Медианная продолжительность жизни = 78.77 лет
- Максимальная продолжительность жизни = 83.27 года

In [227]:
lifeexp[(lifeexp['CountryDisplay'] == 'USA') &
     (lifeexp['GhoDisplay'] == 'Life expectancy at birth (years)')].groupby('CountryDisplay')['Numeric'].agg(['min', 'mean', 'median', 'max'])

min       mean   median       max
CountryDisplay                                    
USA             72.0  77.443821  77.8973  81.13326

Среди статей и продолжительности жизни людей от рождения **в США** выявлены такие показатели, как:

- Минимальная продолжительность жизни = 72 года
- Средняя продолжительность жизни = 77.44 лет
- Медианная продолжительность жизни = 77.9 лет
- Максимальная продолжительность жизни = 81 год

In [228]:
lifeexp[(lifeexp['CountryDisplay'] == 'Russian Federation') &
     (lifeexp['GhoDisplay'] == 'Life expectancy at birth (years)')].groupby('CountryDisplay')['Numeric'].agg(['min', 'mean', 'median', 'max'])

min       mean  median       max
CountryDisplay                                       
Russian Federation  59.0  68.078784    69.0  75.22597

Среди статей и продолжительности жизни людей от рождения **в России** выявлены такие показатели, как:

- Минимальная продолжительность жизни = 59 лет
- Средняя продолжительность жизни = 68 лет
- Медианная продолжительность жизни = 69 лет
- Максимальная продолжительность жизни = 75 лет

### 7

Сформируйте результирующий датафрейм на основе проделанных вами действий в предыдущем номере. У датафрейма должно быть 4 столбца, в соответствии с заданными агреггирующими функциями: минимум, среднее, медиана, максимум, — а также 4 строки: статистика по всему датасету; статистика по трем выбранным вами странам. Сохраните программно созданный вами датафрейм в виде csv файла.

*Примечание. Датафрейм необходимо создавать автоматизированным образом. Не допускается ручной перенос значений из полученных в предыдущем задании ячеек в пустой датафрейм и/или прописывание конкретных значений (чисел) в коде программы. Представьте, что у нас не 3 страны, а 100 и не 4 агреггирующих метрики, а 100!*

In [229]:
full_stats = lifeexp['Numeric'].agg(['min', 'mean', 'median', 'max']).to_frame().T
full_stats.index = ['Overall']

favourite_countries = ['Germany', 'USA', 'Russian Federation']
favourite_stats = lifeexp[(lifeexp['CountryDisplay'].isin(favourite_countries)) & (lifeexp['GhoDisplay'] == 'Life expectancy at birth (years)')].groupby('CountryDisplay')['Numeric'].agg(['min', 'mean', 'median', 'max'])

res_df = pd.concat([full_stats, favourite_stats])

res_df.to_csv('fav_contrs_and_ovrl.csv')
res_df

min       mean    median       max
Overall             11.0  48.145528  56.97971  87.15457
Germany             72.0  78.814174  78.77455  83.27878
Russian Federation  59.0  68.078784  69.00000  75.22597
USA                 72.0  77.443821  77.89730  81.13326

### 8

Есть ли в наших датасетах пропуски? Если да, то в каких столбцах? Сколько пропусков в каждом из них?

Подумайте, как стоит избавиться от пропусков в каждом из столбцов, и сделайте это. Аргументируйте ваши действия.

Проверьте, что пропусков нигде не осталось.

*Примечание. Не забудьте сохранить все изменения! В последующих заданиях вы должны работать с уже обновленными датасетами.*

In [230]:
lifeexp.isnull().sum()

index                               0
GhoCode                             0
GhoDisplay                          0
PublishStateCode                    0
PublishStateDisplay                 0
YearCode                            0
YearDisplay                         0
RegionCode                          0
RegionDisplay                       0
WorldBankIncomeGroupGroupCode    1164
WorldBankIncomeGroupDisplay      1164
CountryCode                         0
CountryDisplay                      0
SexCode                             0
SexDisplay                          0
DisplayValue                        0
Numeric                             0
dtype: int64

В нашем датасете пропуски встретились только в двух столбцах, которые по смыслу одинаковые. Поэтому можно сказать, что всего у нас 1164 пропуска.

Так как основная цель нашего анализа заключается в поиске зависимостей между продолжительностью жизни и объемами потребляемого алкоголя, переменная, отвечающая за уровень дохода имеет, так сказать, второстепенную важность. Поэтому наличие пустых значений именно в этих столбцах не вынуждает нас удалять строки, где есть пустые значения. В связи со всем вышесказанным мы можем не запариваться и просто **заменить пустые значения на "Unknown"**. (если бы данный признак был одним из ключевых, то, конечно же, во избежание искажения данных мы бы заменяли пустые значения медианным значением или модой)

In [231]:
lifeexp['WorldBankIncomeGroupGroupCode'].fillna('Unknown', inplace=True)
lifeexp['WorldBankIncomeGroupDisplay'].fillna('Unknown', inplace=True)

C:\Users\D\AppData\Local\Temp\ipykernel_13616\1174964139.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  lifeexp['WorldBankIncomeGroupGroupCode'].fillna('Unknown', inplace=True)
C:\Users\D\AppData\Local\Temp\ipykernel_13616\1174964139.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as 

In [232]:
lifeexp.isnull().sum().sum()

np.int64(0)

Видим, что пропусков действительно не осталось

### 9

Все ли признаки несут в себе полезную информацию? Почему вы так считаете? Избавьтесь от ненужных столбцов. Свой выбор аргументируйте

*Примечание. Не забудьте сохранить все изменения! В последующих заданиях вы должны работать с уже обновленными датасетами.*

In [233]:
columns_to_drop = ['GhoCode', 'PublishStateCode', 'YearCode', 'RegionCode', 'WorldBankIncomeGroupGroupCode', 'CountryCode', 'SexCode', 'PublishStateDisplay', 'DisplayValue', 'YearDisplay', 'WorldBankIncomeGroupDisplay']
lifeexp = lifeexp.drop(columns=columns_to_drop)

- GhoCode, PublishStateCode, YearCode, RegionCode, WorldBankIncomeGroupGroupCode, CountryCode, SexCode: эти столбцы являются кодовыми значениями данных представленных в соответствующих столбцах Display и дублируют информацию.
- Статус публикации абсолютно неважен в рамках цели нашего исследования
- DisplayValue - целая часть от представленных значений в столбце Numeric. Для получения наиболее точных ответов стоит учитывать значения продолжительности жизни, учитывая дробную часть.
- YearDisplay в рамках цели нашего анализа не требуется учитывать год, в который была опубликована статья
- WorldBankIncomeGroupDisplay в рамках нашего анализа не требуется учитывать уровень дохода населения в странах

In [234]:
lifeexp

index                                       GhoDisplay  \
0         0                 Life expectancy at birth (years)   
1         1  Healthy life expectancy (HALE) at birth (years)   
2         2                 Life expectancy at birth (years)   
3         3                 Life expectancy at birth (years)   
4         4                Life expectancy at age 60 (years)   
...     ...                                              ...   
6403   6403                 Life expectancy at birth (years)   
6404   6404                Life expectancy at age 60 (years)   
6405   6405                 Life expectancy at birth (years)   
6406   6406  Healthy life expectancy (HALE) at birth (years)   
6407   6407                 Life expectancy at birth (years)   

              RegionDisplay CountryDisplay  SexDisplay   Numeric  
0                  Americas      Nicaragua        Male  68.00000  
1                    Europe        Ireland        Male  69.30000  
2     Eastern Mediterranean          Yemen  Both sexes  61.00000  
3                    Africa        Nigeria      Female  48.00000  
4           South_East Asia       Thailand      Female  19.00000  
...                     ...            ...         ...       ...  
6403  Eastern Mediterranean        Morocco  Both sexes  68.00000  
6404        Western Pacific          Palau  Both sexes  16.00000  
6405                 Africa   Burkina Faso      Female  51.00000  
6406                 Europe        Albania        Male  63.50000  
6407                 Europe        Armenia  Both sexes  71.02699  

[6408 rows x 6 columns]

По итогам получили вот такой "чистый" датасет без лишних для данного исследования данных

### 10

Посчитайте, насколько сильно коррелируют друг с другом потребления разных видов алкоголя (beer_servings,	spirit_servings и	wine_servings). Создайте датафрейм с попарными корреляциями этих столбцов друг с другом. Какие выводы можно сделать?

In [235]:
correlations = drinks[['beer_servings', 'spirit_servings', 'wine_servings']].corr()

correlations

beer_servings  spirit_servings  wine_servings
beer_servings         1.000000         0.458819       0.527172
spirit_servings       0.458819         1.000000       0.194797
wine_servings         0.527172         0.194797       1.000000

Потребление пива и вина коррелирует с коэффициентом 0.527, что указывает на умеренную положительную связь. Это может означать, что в странах с высоким потреблением пива часто также наблюдается и высокое потребление вина.

Потребление пива и спиртных напитков коррелирует с коэффициентом 0.459, что также представляет собой умеренную положительную корреляцию. Это указывает на то, что в странах, где люди много пьют пива, они также склонны употреблять и спиртные напитки, хотя связь не так сильна, как между пивом и вином.

Потребление вина и спиртных напитков коррелирует с коэффициентом 0.195, что является довольно слабой положительной связью. Это говорит о том, что в странах с высоким потреблением вина не обязательно будет высокое потребление спиртных напитков и наоборот.

### 11

Каким образом следует осуществить соединение двух имеющихся у нас таблиц? Почему? Соедините таблицы!

Получилось ли у вас задействовать все строки, и если нет, то почему? Что делать с теми строками, которые не удается задействовать, если таковые есть? Ответ обоснуйте.

*Подсказка. В ответе на первый вопрос задания, а также в обосновании способа соединения таблиц вам может помочь задание 3 данного раздела ДЗ и соответствующий семинарский ноутбук!*

*Примечание. Не забудьте сохранить все изменения! В последующих заданиях вы должны работать с одним объединенным датасетом.*

In [236]:
merged_df = pd.merge(lifeexp, drinks, how='inner', left_on='CountryDisplay', right_on='country')
merged_df

index_x                                       GhoDisplay  \
0           0                 Life expectancy at birth (years)   
1           1  Healthy life expectancy (HALE) at birth (years)   
2           2                 Life expectancy at birth (years)   
3           3                 Life expectancy at birth (years)   
4           4                Life expectancy at age 60 (years)   
...       ...                                              ...   
6364     6403                 Life expectancy at birth (years)   
6365     6404                Life expectancy at age 60 (years)   
6366     6405                 Life expectancy at birth (years)   
6367     6406  Healthy life expectancy (HALE) at birth (years)   
6368     6407                 Life expectancy at birth (years)   

              RegionDisplay CountryDisplay  SexDisplay   Numeric  index_y  \
0                  Americas      Nicaragua        Male  68.00000      122   
1                    Europe        Ireland        Male  69.30000       81   
2     Eastern Mediterranean          Yemen  Both sexes  61.00000      190   
3                    Africa        Nigeria      Female  48.00000      124   
4           South_East Asia       Thailand      Female  19.00000      169   
...                     ...            ...         ...       ...      ...   
6364  Eastern Mediterranean        Morocco  Both sexes  68.00000      114   
6365        Western Pacific          Palau  Both sexes  16.00000      129   
6366                 Africa   Burkina Faso      Female  51.00000       26   
6367                 Europe        Albania        Male  63.50000        1   
6368                 Europe        Armenia  Both sexes  71.02699        7   

           country  beer_servings  spirit_servings  wine_servings  \
0        Nicaragua             78              118              1   
1          Ireland            313              118            165   
2            Yemen              6                0              0   
3          Nigeria             42                5              2   
4         Thailand             99              258              1   
...            ...            ...              ...            ...   
6364       Morocco             12                6             10   
6365         Palau            306               63             23   
6366  Burkina Faso             25                7              7   
6367       Albania             89              132             54   
6368       Armenia             21              179             11   

      total_litres_of_pure_alcohol  
0                              3.5  
1                             11.4  
2                              0.1  
3                              9.1  
4                              6.4  
...                            ...  
6364                           0.5  
6365                           6.9  
6366                           4.3  
6367                           4.9  
6368                           3.8  

[6369 rows x 12 columns]

In [237]:
merged_df.CountryDisplay.nunique()

193

- Начнем с того, что в датасете дринкс всего 194 уникальные страны, а в датасете о продолжительности жизни - 193. Нам точно необходимо соединить эти две таблицы пересечением, так как, как уже много раз повторялось, мы исследуем зависимость продолжительности жизни от объемов потребляемого алкоголя, то есть, нам никакой информации не дадут те строки, в которых есть либо только данные о потребляемом алкоголе (в определенной стране), либо данные только о продолжительности жизни (в определенной стране). Следовательно, нам следует оставить только те строки, которые отвечают требованиям цели нашего исследования - те, в которых есть данные и о продолжительности жизни, и о потребляемом алкоголе (в определенной стране).
- После соединения таблиц мы имеем данные о 193 уникальных странах

### 12

Осуществите группировку данных по странам и типу научной статьи (многомерная группировка). В качестве столбцов значений для группировки возьмите оцененную продолжительность жизни и общее потребление алкоголя в литрах. В качестве агреггирующих функций для продолжительности жизни возьмите среднее, медиану, минимум и максимум, а для потребления алкоголя — агреггирующую функцию/функции выберите самостоятельно. Как вы можете прокомментировать свой выбор и что вы наблюдаете?

In [238]:
grouped_df = merged_df.groupby(['CountryDisplay', 'GhoDisplay']).agg({'Numeric': ['mean', 'median', 'min', 'max'],
                                                                      'total_litres_of_pure_alcohol': ['mean', 'sum']})
grouped_df

Numeric  \
                                                                     mean   
CountryDisplay GhoDisplay                                                   
Afghanistan    Healthy life expectancy (HALE) at birth (years)  47.833333   
               Life expectancy at age 60 (years)                15.223012   
               Life expectancy at birth (years)                 56.311805   
Albania        Healthy life expectancy (HALE) at birth (years)  63.800000   
               Life expectancy at age 60 (years)                17.728610   
...                                                                   ...   
Zambia         Life expectancy at age 60 (years)                16.068835   
               Life expectancy at birth (years)                 49.953199   
Zimbabwe       Healthy life expectancy (HALE) at birth (years)  45.222222   
               Life expectancy at age 60 (years)                17.369381   
               Life expectancy at birth (years)                 55.186778   

                                                                           \
                                                                   median   
CountryDisplay GhoDisplay                                                   
Afghanistan    Healthy life expectancy (HALE) at birth (years)  48.800000   
               Life expectancy at age 60 (years)                15.061535   
               Life expectancy at birth (years)                 57.000000   
Albania        Healthy life expectancy (HALE) at birth (years)  63.900000   
               Life expectancy at age 60 (years)                18.000000   
...                                                                   ...   
Zambia         Life expectancy at age 60 (years)                16.000000   
               Life expectancy at birth (years)                 51.000000   
Zimbabwe       Healthy life expectancy (HALE) at birth (years)  47.700000   
               Life expectancy at age 60 (years)                17.654040   
               Life expectancy at birth (years)                 58.363385   

                                                                      \
                                                                 min   
CountryDisplay GhoDisplay                                              
Afghanistan    Healthy life expectancy (HALE) at birth (years)  44.6   
               Life expectancy at age 60 (years)                13.0   
               Life expectancy at birth (years)                 49.0   
Albania        Healthy life expectancy (HALE) at birth (years)  59.9   
               Life expectancy at age 60 (years)                15.0   
...                                                              ...   
Zambia         Life expectancy at age 60 (years)                14.0   
               Life expectancy at birth (years)                 40.0   
Zimbabwe       Healthy life expectancy (HALE) at birth (years)  36.4   
               Life expectancy at age 60 (years)                15.0   
               Life expectancy at birth (years)                 42.0   

                                                                          \
                                                                     max   
CountryDisplay GhoDisplay                                                  
Afghanistan    Healthy life expectancy (HALE) at birth (years)  50.50000   
               Life expectancy at age 60 (years)                17.00000   
               Life expectancy at birth (years)                 61.97733   
Albania        Healthy life expectancy (HALE) at birth (years)  66.00000   
               Life expectancy at age 60 (years)                20.00000   
...                                                                  ...   
Zambia         Life expectancy at age 60 (years)                18.10310   
               Life expectancy at birth (years)                 60.01617   
Zimbabwe       Healthy life expectancy (HALE) at birth (years) 

Для потребления алкоголя я выбрал среднее и сумму. Среднее значение показывает общий уровень потребления алкоголя на душу населения, а сумма может дать представление о совокупном потреблении в рамках каждой категории статей, что будет полезно для примерной оценки воздействия алкоголя на население.

### 13

Вставьте любые два столбца из предыдущего задания, которые вы получили путем группировки, в исходную таблицу в качестве двух новых столбцов (назвать вы их можете как хотите). Разумеется, значения этих столбцов должны соответствовать значениям исходной таблицы по указанным в прошлом задании категориям.

*Примечание. Не забудьте сохранить все изменения! В последующих заданиях вы должны работать с обновленным датасетом.*

In [239]:
merged_df['life_exp_mean'] = merged_df.groupby(['CountryDisplay', 'GhoDisplay'])['Numeric'].transform('mean')
merged_df['alc_consump_mean'] = merged_df.groupby(['CountryDisplay'])['total_litres_of_pure_alcohol'].transform('mean')
merged_df

index_x                                       GhoDisplay  \
0           0                 Life expectancy at birth (years)   
1           1  Healthy life expectancy (HALE) at birth (years)   
2           2                 Life expectancy at birth (years)   
3           3                 Life expectancy at birth (years)   
4           4                Life expectancy at age 60 (years)   
...       ...                                              ...   
6364     6403                 Life expectancy at birth (years)   
6365     6404                Life expectancy at age 60 (years)   
6366     6405                 Life expectancy at birth (years)   
6367     6406  Healthy life expectancy (HALE) at birth (years)   
6368     6407                 Life expectancy at birth (years)   

              RegionDisplay CountryDisplay  SexDisplay   Numeric  index_y  \
0                  Americas      Nicaragua        Male  68.00000      122   
1                    Europe        Ireland        Male  69.30000       81   
2     Eastern Mediterranean          Yemen  Both sexes  61.00000      190   
3                    Africa        Nigeria      Female  48.00000      124   
4           South_East Asia       Thailand      Female  19.00000      169   
...                     ...            ...         ...       ...      ...   
6364  Eastern Mediterranean        Morocco  Both sexes  68.00000      114   
6365        Western Pacific          Palau  Both sexes  16.00000      129   
6366                 Africa   Burkina Faso      Female  51.00000       26   
6367                 Europe        Albania        Male  63.50000        1   
6368                 Europe        Armenia  Both sexes  71.02699        7   

           country  beer_servings  spirit_servings  wine_servings  \
0        Nicaragua             78              118              1   
1          Ireland            313              118            165   
2            Yemen              6                0              0   
3          Nigeria             42                5              2   
4         Thailand             99              258              1   
...            ...            ...              ...            ...   
6364       Morocco             12                6             10   
6365         Palau            306               63             23   
6366  Burkina Faso             25                7              7   
6367       Albania             89              132             54   
6368       Armenia             21              179             11   

      total_litres_of_pure_alcohol  life_exp_mean  alc_consump_mean  
0                              3.5      72.667402               3.5  
1                             11.4      69.911111              11.4  
2                              0.1      61.567050               0.1  
3                              9.1      50.473338               9.1  
4                              6.4      19.895824               6.4  
...                            ...            ...               ...  
6364                           0.5      68.574297               0.5  
6365                           6.9      17.313487               6.9  
6366                           4.3      53.880190               4.3  
6367                           4.9      63.800000               4.9  
6368                           3.8      70.046697               3.8  

[6369 rows x 14 columns]

### 14

Используя группировки, аргументированно ответьте на вопрос: на каких континентах средняя ожидаемая продолжительность жизни выше у мужчин, а на каких у женщин?

А медианная ожидаемая продолжительность?

А минимальная ожидаемая продолжительность?

Сделайте выводы!

*Подсказка. Казалось бы, при чем здесь тип научной статьи?*

In [240]:
life_exp_at_birth_stats = merged_df[merged_df['GhoDisplay'] == 'Life expectancy at birth (years)'].groupby(['RegionDisplay', 'SexDisplay']).agg({
    'Numeric': ['mean', 'median', 'min']}).reset_index()
life_exp_at_birth_stats

RegionDisplay  SexDisplay    Numeric                 
                                            mean     median   min
0                  Africa  Both sexes  56.244803  57.000000  38.0
1                  Africa      Female  57.847483  58.000000  38.0
2                  Africa        Male  54.720970  55.000000  38.0
3                Americas  Both sexes  73.028041  74.000000  54.0
4                Americas      Female  75.873706  77.000000  56.0
5                Americas        Male  70.282002  71.000000  52.0
6   Eastern Mediterranean  Both sexes  68.828298  70.989760  47.0
7   Eastern Mediterranean      Female  70.590926  73.000000  50.0
8   Eastern Mediterranean        Male  67.283447  69.000000  45.0
9                  Europe  Both sexes  75.233666  76.000000  62.0
10                 Europe      Female  78.454599  79.000000  65.0
11                 Europe        Male  71.976935  72.518745  58.0
12        South_East Asia  Both sexes  66.345093  67.500000  50.0
13        South_East Asia      Female  68.188877  68.968250  51.0
14        South_East Asia        Male  64.491738  65.000000  48.0
15        Western Pacific  Both sexes  70.985250  71.000000  53.0
16        Western Pacific      Female  73.538356  73.500000  54.0
17        Western Pacific        Male  68.585153  68.124050  51.0

- Получается так, что средняя ожидаемая продолжительность жизни на всех континентах выше у женщин, чем у мужчин, также как и медианная. Что касается минимальной продолжительности жизни, все не так однозначно, но все же общая тенденция более высокой продолжительности жизни у женщин сохраняется.
- Для получения результатов этого мини-анализа были взяты только статьи "Life expectancy at birth (years)", так как нам требуется выявить именно продолжительность жизни, то есть жизнь должна быть взята от рождения.

### 15

Работая лишь с общим объединенным датасетом и не используя исходную таблицу drinks, подсчитайте среднее потребление каждого вида алкоголя и среднее общее количество литров потребляемого алкоголя на человека по всем континентам. Занесите ваши результаты в отдельный датафрейм!

*Подсказка. Задача немного сложнее, чем может показаться на первый взгляд — в объединенном датасете страны встречаются разное количество раз, и из-за этого могут возникнуть определенные проблемы. Если так случится, надо подумать, как это обойти!*

*Примечание. Разумеется, это необходимо сделать автоматизированно, а не перебирая каждый континент по отдельности.*

In [241]:
unique_countries_alcohol_consumption = merged_df[['CountryDisplay', 'RegionDisplay', 'beer_servings', 'spirit_servings', 'wine_servings', 'total_litres_of_pure_alcohol']].drop_duplicates(subset='CountryDisplay')
continent_alcohol_consumption_avg = unique_countries_alcohol_consumption.groupby('RegionDisplay').mean(numeric_only=True)
continent_alcohol_consumption_avg

beer_servings  spirit_servings  wine_servings  \
RegionDisplay                                                          
Africa                     68.826087        17.304348      18.000000   
Americas                  155.600000       148.257143      37.514286   
Eastern Mediterranean      10.428571        21.190476       4.857143   
Europe                    175.132075       130.415094     123.528302   
South_East Asia            14.818182        44.090909       0.454545   
Western Pacific            81.962963        66.851852      27.740741   

                       total_litres_of_pure_alcohol  
RegionDisplay                                        
Africa                                     3.360870  
Americas                                   6.102857  
Eastern Mediterranean                      0.714286  
Europe                                     7.873585  
South_East Asia                            1.063636  
Western Pacific                            3.637037

### 16

Создание новых признаков (feature engineering) является одним из основных средств улучшения качества работы алгоритмов машинного обучения на этапе обработки данных, а также отличный способ увеличить глубину ваших выводов из проделанного анализа в целом! Частично мы с вами уже осуществляли feature engineering, когда добавляли новые столбцы в задании 13 данного раздела.

Теперь давайте добавим в таблицу столбец, который будет показывать, какое общее количество servings per person производится в среднем по стране, к которой относится научная статья. Столбец для этого признака назовите 'total_servings'.

Чем может быть полезен такой признак?

***Бонус (2 бон. балла):*** Придумайте еще какой-нибудь свой осмысленный признак, который может быть полезен для анализа (несет в себе какую-то информацию), и добавьте его в таблицу. Обязательно аргументируйте, чем полезен ваш признак.

In [242]:
merged_df['total_servings'] = merged_df['beer_servings'] + merged_df['spirit_servings'] + merged_df['wine_servings']
merged_df

index_x                                       GhoDisplay  \
0           0                 Life expectancy at birth (years)   
1           1  Healthy life expectancy (HALE) at birth (years)   
2           2                 Life expectancy at birth (years)   
3           3                 Life expectancy at birth (years)   
4           4                Life expectancy at age 60 (years)   
...       ...                                              ...   
6364     6403                 Life expectancy at birth (years)   
6365     6404                Life expectancy at age 60 (years)   
6366     6405                 Life expectancy at birth (years)   
6367     6406  Healthy life expectancy (HALE) at birth (years)   
6368     6407                 Life expectancy at birth (years)   

              RegionDisplay CountryDisplay  SexDisplay   Numeric  index_y  \
0                  Americas      Nicaragua        Male  68.00000      122   
1                    Europe        Ireland        Male  69.30000       81   
2     Eastern Mediterranean          Yemen  Both sexes  61.00000      190   
3                    Africa        Nigeria      Female  48.00000      124   
4           South_East Asia       Thailand      Female  19.00000      169   
...                     ...            ...         ...       ...      ...   
6364  Eastern Mediterranean        Morocco  Both sexes  68.00000      114   
6365        Western Pacific          Palau  Both sexes  16.00000      129   
6366                 Africa   Burkina Faso      Female  51.00000       26   
6367                 Europe        Albania        Male  63.50000        1   
6368                 Europe        Armenia  Both sexes  71.02699        7   

           country  beer_servings  spirit_servings  wine_servings  \
0        Nicaragua             78              118              1   
1          Ireland            313              118            165   
2            Yemen              6                0              0   
3          Nigeria             42                5              2   
4         Thailand             99              258              1   
...            ...            ...              ...            ...   
6364       Morocco             12                6             10   
6365         Palau            306               63             23   
6366  Burkina Faso             25                7              7   
6367       Albania             89              132             54   
6368       Armenia             21              179             11   

      total_litres_of_pure_alcohol  life_exp_mean  alc_consump_mean  \
0                              3.5      72.667402               3.5   
1                             11.4      69.911111              11.4   
2                              0.1      61.567050               0.1   
3                              9.1      50.473338               9.1   
4                              6.4      19.895824               6.4   
...                            ...            ...               ...   
6364                           0.5      68.574297               0.5   
6365                           6.9      17.313487               6.9   
6366                           4.3      53.880190               4.3   
6367                           4.9      63.800000               4.9   
6368                           3.8      70.046697               3.8   

      total_servings  
0                197  
1                596  
2                  6  
3                 49  
4                358  
...              ...  
6364              28  
6365             392  
6366              39  
6367             275  
6368             211  

[6369 rows x 15 columns]

- Столбец total_servings представляет собой агрегированный показатель, который может быть полезен для оценки общего уровня потребления алкоголя в разных странах.
- Наличие этого показателя в датасете позволяет провести более глубокий анализ взаимосвязи между уровнем потребления алкоголя и другими исследуемыми переменными, такими как продолжительность жизни.

In [243]:
merged_df['avg_alcohol_per_serving'] = merged_df['total_litres_of_pure_alcohol'] / merged_df['total_servings']
merged_df

index_x                                       GhoDisplay  \
0           0                 Life expectancy at birth (years)   
1           1  Healthy life expectancy (HALE) at birth (years)   
2           2                 Life expectancy at birth (years)   
3           3                 Life expectancy at birth (years)   
4           4                Life expectancy at age 60 (years)   
...       ...                                              ...   
6364     6403                 Life expectancy at birth (years)   
6365     6404                Life expectancy at age 60 (years)   
6366     6405                 Life expectancy at birth (years)   
6367     6406  Healthy life expectancy (HALE) at birth (years)   
6368     6407                 Life expectancy at birth (years)   

              RegionDisplay CountryDisplay  SexDisplay   Numeric  index_y  \
0                  Americas      Nicaragua        Male  68.00000      122   
1                    Europe        Ireland        Male  69.30000       81   
2     Eastern Mediterranean          Yemen  Both sexes  61.00000      190   
3                    Africa        Nigeria      Female  48.00000      124   
4           South_East Asia       Thailand      Female  19.00000      169   
...                     ...            ...         ...       ...      ...   
6364  Eastern Mediterranean        Morocco  Both sexes  68.00000      114   
6365        Western Pacific          Palau  Both sexes  16.00000      129   
6366                 Africa   Burkina Faso      Female  51.00000       26   
6367                 Europe        Albania        Male  63.50000        1   
6368                 Europe        Armenia  Both sexes  71.02699        7   

           country  beer_servings  spirit_servings  wine_servings  \
0        Nicaragua             78              118              1   
1          Ireland            313              118            165   
2            Yemen              6                0              0   
3          Nigeria             42                5              2   
4         Thailand             99              258              1   
...            ...            ...              ...            ...   
6364       Morocco             12                6             10   
6365         Palau            306               63             23   
6366  Burkina Faso             25                7              7   
6367       Albania             89              132             54   
6368       Armenia             21              179             11   

      total_litres_of_pure_alcohol  life_exp_mean  alc_consump_mean  \
0                              3.5      72.667402               3.5   
1                             11.4      69.911111              11.4   
2                              0.1      61.567050               0.1   
3                              9.1      50.473338               9.1   
4                              6.4      19.895824               6.4   
...                            ...            ...               ...   
6364                           0.5      68.574297               0.5   
6365                           6.9      17.313487               6.9   
6366                           4.3      53.880190               4.3   
6367                           4.9      63.800000               4.9   
6368                           3.8      70.046697               3.8   

      total_servings  avg_alcohol_per_serving  
0                197                 0.017766  
1                596                 0.019128  
2                  6                 0.016667  
3                 49                 0.185714  
4                358                 0.017877  
...              ...                      ...  
6364              28                 0.017857  
6365             392                 0.017602  
6366              39                 0.110256  
6367             275                 0.017818  
6368             211                 0.018009  

[6369 rows x 16 columns]

Я решил добавить признак, описывающий какое количество чистого алкоголя в среднем приходится на одну порцию. Это может быть полезно при оценке концентрации алкоголя: например в одной стране очень много порций, но и люди там употребляют в основном пиво - концентрация будет низкой. Также по одному лишь этому признаку можно сравнить две страны и сказать, где потребление алкоголя слабее или крепче.

### 17

Есть ли в данных категориальные признаки, которые принимают значения строк, а не чисел? Если да, то для многих алгоритмов это может стать серьезным препятствием, поэтому закодируйте категориальные признаки так, чтобы к данным можно было применить любые алгоритмы. Аргументируйте выбор методов кодирования.

*Подсказка. Вам могут помочь Label encoding и One-Hot Encoding. Но будьте внимательны! У каждого из этих методов есть свои сильные и слабые стороны, а также ограничения применимости.*

In [244]:
merged_df['SexDisplay'] = merged_df['SexDisplay'].replace({'Male': 1, 'Female': 2, 'Both sexes': 3})
merged_df['SexDisplay'] = merged_df['SexDisplay'].replace({'Life expectancy at birth (years)': 1,
                                                           'Life expectancy at age 60 (years)': 2,
                                                           'Healthy life expectancy (HALE) at birth (years)': 3})

C:\Users\D\AppData\Local\Temp\ipykernel_13616\4187025585.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  merged_df['SexDisplay'] = merged_df['SexDisplay'].replace({'Male': 1, 'Female': 2, 'Both sexes': 3})


Эти два признака принимают ограниченное количество значений и могут быть легко представлены числами без потери информации.

### 18

Придумайте какое-нибудь своё интересное задание на анализ предлагаемых данных и решите его.

Креативность в этом задании приветствуется, и за нее можно получить ***1-2 бонусных балла***!

*Подсказка. Вы можете вдохновляться различными заданиями из этого раздела ДЗ, как, например, заданиями: 4, 6, 10, 12, 14, 15, 16. Но, чтобы получить бонусные баллы, лучше не используйте аналогичные идеи, а придумайте что-то своё!*

In [245]:
# your code here

## Бонусная часть на анализ целевой переменной (до 5 бонусных баллов)

После всего проделанного анализа и подготовки данных мы можем перейти непосредственно к ответу на исходный вопрос исследования, который был сформулирован вами в задании 3 данного раздела.

### 19

***Бонус (5 бон. баллов):*** Дайте развернутый, обоснованный, подкрепленный данными и написанным вами кодом ответ на вопрос, влияют ли признаки зависимой таблицы на целевую переменную из главной таблицы и, если да, то каким именно образом влияют?

Обязательно аргументируйте полученные вами результаты и выводы. Чем подробнее и обоснованнее анализ вы проведете в рамках данного бонусного задания, тем больше бонусных баллов получите! Вы можете использовать подходы и методы из рассмотренных на семинаре и в рамках данного домашнего задания, а можете придумать и что-то своё. Креативность максимально приветствуется!

In [246]:
# your code here

    {your final conclusion here}